### PR TITLE
Lint TensorFlow: Use TPUs

### PR DESCRIPTION
**Pull request by 8bitmp3**
Splitting PR #1736 into smaller digestible bits. 🙏 @lamberta @MarkDaoust

### CODE REVIEW

**Comment by 8bitmp3** [comment]
If you have time @lamberta @MarkDaoust PTAL, thanks!

File: site/en/guide/tpu.ipynb [hunk 1]
```diff
@@ -61,10 +61,7 @@
         "id": "Ys81cOhXOWUP"
       },
       "source": [
-        "Experimental support for Cloud TPUs is currently available for Keras\n",
-        "and Google Colab. Before you run this Colab notebooks, ensure that\n",
-        "your hardware accelerator is a TPU by checking your notebook settings:\n",
-        "Runtime > Change runtime type > Hardware accelerator > TPU."
+        "Experimental support for Cloud TPUs is currently available for Keras and [Google Colaboratory (Colab)](https://colab.research.google.com). Before you run this Colab notebook, make sure that your hardware accelerator is a TPU by checking your notebook settings: **Runtime** > **Change runtime type** > **Hardware accelerator** > **TPU**."
```
Comment (lamberta) [inline site/en/guide/tpu.ipynb:64]:
```suggestion
        "Experimental support for Cloud TPUs is currently available for Keras and Google Colab. Before you run this Colab notebook, make sure that your hardware accelerator is a TPU by checking your notebook settings: **Runtime** > **Change runtime type** > **Hardware accelerator** > **TPU**."
```

By the time users get to this doc, we don't need to keep introducing Colab.

File: site/en/guide/tpu.ipynb [hunk 2]
```diff
@@ -96,8 +93,9 @@
         "id": "yDWaRxSpwBN1"
       },
       "source": [
-        "## TPU Initialization\n",
-        "TPUs are usually on Cloud TPU workers which are different from the local process running the user python program. Thus some initialization work needs to be done to connect to the remote cluster and initialize TPUs. Note that the `tpu` argument to `TPUClusterResolver` is a special address just for Colab. In the case that you are running on Google Compute Engine (GCE), you should instead pass in the name of your CloudTPU."
+        "## TPU initialization\n",
+        "\n",
+        "TPUs are typically Cloud TPU workers, which are different from the local process running the user's Python program. Thus, you need to do some initialization work to connect to the remote cluster and initialize the TPUs. Note that the `tpu` argument to `TPUClusterResolver` is a special address just for Colab. If you are running your code on Google Compute Engine (GCE), you should instead pass in the name of your Cloud TPU."
```
Comment (lamberta) [inline site/en/guide/tpu.ipynb:98]: while you're here, can you add the full symbol name for `TPUClusterResolver` so we get the auto-link behavior?

File: site/en/guide/tpu.ipynb [hunk 3]
```diff
@@ -157,7 +158,8 @@
       },
       "source": [
         "## Distribution strategies\n",
-        "Most times users want to run the model on multiple TPUs in a data parallel way. A distribution strategy is an abstraction that can be used to drive models on CPU, GPUs or TPUs. Simply swap out the distribution strategy and the model will run on the given device. See the [distribution strategy guide](./distributed_training.ipynb) for more information."
+        "\n",
+        "Usually, you would want to run your model on multiple TPUs in a data-parallel way. To distribute your model on multiple TPUs (or other accelerators), TensorFlow offers several distribution strategies. You can replace your distribution strategy and the model will run on any given (TPU) device. Check the [distribution strategy guide](./distributed_training.ipynb) for more information."
```
Comment (lamberta) [inline site/en/guide/tpu.ipynb:162]:
```suggestion
        "Usually you run your model on multiple TPUs in a data-parallel way. To distribute your model on multiple TPUs (or other accelerators), TensorFlow offers several distribution strategies. You can replace your distribution strategy and the model will run on any given (TPU) device. Check the [distribution strategy guide](./distributed_training.ipynb) for more information."
```
First sentence is a bit awkward and can tighten up

File: site/en/guide/tpu.ipynb [hunk 4]
```diff
@@ -166,7 +168,7 @@
         "id": "DcDPMZs-9uLJ"
       },
       "source": [
-        "First, creates the `TPUStrategy` object."
+        "To demonstrate this, create a `TPUStrategy` object:"
```
Comment (lamberta) [inline site/en/guide/tpu.ipynb:171]: Use full symbol name for `TPUStrategy`

File: site/en/guide/tpu.ipynb [hunk 5]
```diff
@@ -250,18 +254,19 @@
...
       },
       "source": [
-        "### Input datasets\n",
-        "Efficient use of the `tf.data.Dataset` API is critical when using a Cloud TPU, as it is impossible to use the Cloud TPUs unless you can feed them data quickly enough. See [Input Pipeline Performance Guide](./data_performance.ipynb) for details on dataset performance.\n",
+        "### Load the dataset\n",
+        "\n",
+        "Efficient use of the `tf.data.Dataset` API is critical when using a Cloud TPU, as it is impossible to use the Cloud TPUs unless you can feed them data quickly enough. You can learn more about dataset performance in the [Input pipeline performance guide](./data_performance.ipynb).\n",
         "\n",
-        "For all but the simplest experimentation (using `tf.data.Dataset.from_tensor_slices` or other in-graph data) you will need to store all data files read by the Dataset in Google Cloud Storage (GCS) buckets.\n",
+        "For all but the simplest experimentations (using `tf.data.Dataset.from_tensor_slices` or other in-graph data) you will need to store all data files read by the Dataset in Google Cloud Storage (GCS) buckets.\n",
```
Comment (lamberta) [inline site/en/guide/tpu.ipynb:261]:
```suggestion
        "For all but the simplest experiments (using `tf.data.Dataset.from_tensor_slices` or other in-graph data), you need to store all data files read by the Dataset in Google Cloud Storage (GCS) buckets.\n",
```

File: site/en/guide/tpu.ipynb [hunk 6]
```diff
@@ -250,18 +254,19 @@
...
         "\n",
-        "For most use-cases, it is recommended to convert your data into `TFRecord` format and use a `tf.data.TFRecordDataset` to read it. See [TFRecord and tf.Example tutorial](../tutorials/load_data/tfrecord.ipynb) for details on how to do this. This, however, is not a hard requirement and you can use other dataset readers (`FixedLengthRecordDataset` or `TextLineDataset`) if you prefer.\n",
+        "For most use cases, you're recommended to convert your data into the `TFRecord` format and use a `tf.data.TFRecordDataset` to read it. Check the [TFRecord and tf.Example tutorial](../tutorials/load_data/tfrecord.ipynb) for details on how to do this. It is not a hard requirement and you can use other dataset readers, such as `FixedLengthRecordDataset` or `TextLineDataset`.\n",
```
Comment (lamberta) [inline site/en/guide/tpu.ipynb:263]:
```suggestion
        "For most use cases, it is recommended to convert your data into the `TFRecord` format and use a `tf.data.TFRecordDataset` to read it. Check the [TFRecord and tf.Example tutorial](../tutorials/load_data/tfrecord.ipynb) for details on how to do this. It is not a hard requirement and you can use other dataset readers, such as `FixedLengthRecordDataset` or `TextLineDataset`.\n",
```

Please use full symbol names for `FixedLengthRecordDataset` and `TextLineDataset` to auto-link

File: site/en/guide/tpu.ipynb [hunk 7]
```diff
@@ -250,18 +254,19 @@
...
         "\n",
-        "Small datasets can be loaded entirely into memory using `tf.data.Dataset.cache`.\n",
+        "You can load entire small datasets into memory using `tf.data.Dataset.cache`.\n",
         "\n",
-        "Regardless of the data format used, it is strongly recommended that you use large files, on the order of 100MB. This is especially important in this networked setting as the overhead of opening a file is significantly higher.\n",
+        "Regardless of the data format used, you're strongly recommended to use large files on the order of 100MB. This is especially important in this networked setting, as the overhead of opening a file is significantly higher.\n",
```
Comment (lamberta) [inline site/en/guide/tpu.ipynb:267]: I don't think this change is compelling

File: site/en/guide/tpu.ipynb [hunk 8]
```diff
@@ -343,7 +348,7 @@
         "id": "8hSGBIYtUugJ"
       },
       "source": [
-        "To reduce python overhead, and maximize the performance of your TPU, try out the **experimental** `experimental_steps_per_execution` argument to `Model.compile`. Here it increases throughput by about 50%:"
+        "To reduce Python overhead and maximize the performance of your TPU, you can pass in the **experimental** argument—`experimental_steps_per_execution`—to `Model.compile`. In this example, it increases throughput by about 50%:"
```
Comment (lamberta) [inline site/en/guide/tpu.ipynb:351]:
```suggestion
        "To reduce Python overhead and maximize the performance of your TPU, pass in the **experimental** argument—`experimental_steps_per_execution`—to `Model.compile`. In this example, it increases throughput by about 50%:"
```

While "you" is allowed, generally prefer to write so it's not needed. /Implicit you/, I suppose. Repeated use is a bit too friendly, imho.

**Review by lamberta — CHANGES_REQUESTED** [review]
Thanks! Some comments

File: site/en/guide/tpu.ipynb (hunk 7 above)
Comment (8bitmp3) [reply site/en/guide/tpu.ipynb:267]: Gotcha

File: site/en/guide/tpu.ipynb (hunk 6 above)
Comment (8bitmp3) [reply site/en/guide/tpu.ipynb:263]: On it @lamberta 

⬇️

`tf.data.FixedLengthRecordDataset` <- `FixedLengthRecordDataset`
`tf.data.TextLineDataset` <- `TextLineDataset`

File: site/en/guide/tpu.ipynb (hunk 4 above)
Comment (8bitmp3) [reply site/en/guide/tpu.ipynb:171]: 👍

`tf.distribute.TPUStrategy` <-`TPUStrategy`

File: site/en/guide/tpu.ipynb (hunk 5 above)
Comment (8bitmp3) [reply site/en/guide/tpu.ipynb]: done 👍

`tf.distribute.TPUStrategy` <- `TPUStrategy`

**Comment by 8bitmp3** [comment]
This is sort of related @lamberta: whats your view on minor refreshing of the [`tpu_strategy.py`](https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/distribute/tpu_strategy.py) ("master") API docs. For example:

```diff
    Raises:
      ValueError: 1) If the size of partition_dimensions does not equal to rank
-         of `tensor` or 2) if product of elements of `partition_dimensions` does
+         of `tensor`; or 2) if product of elements of `partition_dimensions` does
-         not match the number of logical devices per replica defined by the
+         not match the number of logical devices per replica (defined by the
-         implementing DistributionStrategy's device specification or
+         implementing DistributionStrategy's device specification); or
        3) if a known size of `tensor` is not divisible by corresponding
        value in `partition_dimensions`.
```

Overall, this API looks great and I've added a small fix to enable auto-linking (small PR). Personally, it took a few "passes" to wrap the head around these docs because of the use of complex sentences. (In addition, maybe adding a few definite/indefinite articles like "a" and "the" may help too). Originally, I noticed the challenge of understanding stuff here https://www.tensorflow.org/api_docs/python/tf/distribute/TPUStrategy. But the docs are super detailed 👍 LMKWYT

File: site/en/guide/tpu.ipynb (hunk 2 above)
Comment (8bitmp3) [reply site/en/guide/tpu.ipynb:98]: Done 👍

**Comment by 8bitmp3** [comment]
Hi @lamberta The tests have passed, I believe. PTAL 👍

**Comment by 8bitmp3** [comment]
LGTM

**Comment by 8bitmp3** [comment]
Waiting for the `Merge branch 'tensorflow:master' into lint-tpu-guide` commit to propagate (in the original `lint-tpu-guide` repo -> fetched and merged from upstream `tensorflow:master`).

**Review by lamberta — APPROVED** [review]

**Pull request merged**